File: backend/app/schemas/formal_report.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from app.schemas.trust import KnowledgeCitation

FORMAL_CHAPTER_TITLES = (
    "家庭基础情况",
    "理财目标",
    "大额支出计划",
    "理财假设",
    "家庭财务报表",
    "家庭财务比率分析",
    "投资规划建议",
    "免责声明",
)

ReportStatus = Literal["draft_requires_human_review", "workflow_linked", "client_ready"]
ReportTrigger = Literal["manual", "monthly_review", "major_event", "action_status_change"]
ReportActionStatus = Literal["open", "completed", "deferred", "not_applicable"]


class ReportTable(BaseModel):
    title: str
    columns: list[str] = Field(min_length=1)
    rows: list[list[str]]
    note: str = ""
    calculation_source: str

    @model_validator(mode="after")
    def rows_match_columns(self) -> ReportTable:
        if any(len(row) != len(self.columns) for row in self.rows):
            raise ValueError("报告表格每行必须与列数一致")
        return self


class ReportAdvice(BaseModel):
    code: str
    title: str
    reason: str
    priority: int = Field(ge=1)
    action: str
    completion_criteria: str
    review_cycle: str
    status: ReportActionStatus = "open"
    due_date: date | None = None
    citation_ids: list[str] = Field(default_factory=list)
    calculation_source: str


class ReportSection(BaseModel):
    code: str
    title: str
    narratives: list[str] = Field(default_factory=list)
    tables: list[ReportTable] = Field(default_factory=list)
    advice: list[ReportAdvice] = Field(default_factory=list)
    citation_ids: list[str] = Field(default_factory=list)


class FormalReportChapter(BaseModel):
    number: int = Field(ge=1, le=8)
    title: str
    summary: str
    sections: list[ReportSection] = Field(min_length=1)


class ReportAppendix(BaseModel):
    code: str
    title: str
    tables: list[ReportTable] = Field(default_factory=list)
    narratives: list[str] = Field(default_factory=list)


class ReportSourceClaim(BaseModel):
    text: str
    citation_ids: list[str] = Field(min_length=1)


class ReportNumericClaim(BaseModel):
    code: str
    label: str
    raw_value: str
    display_value: str
    unit: str
    source_path: str
    calculation_source: Literal[
        "deterministic_tools",
        "deterministic_simulation_engine",
        "deterministic_review_schedule",
    ]


class ReportVersionLedger(BaseModel):
    report_version: str
    input_version: str
    formula_version: str
    planning_rule_version: str
    portfolio_rule_version: str
    twin_result_version: str
    model_version: str
    prompt_version: str
    knowledge_version: str
    product_catalog_version: str
    fund_advisory_catalog_version: str | None = None
    workflow_version: str | None


class ReportExecutionMetrics(BaseModel):
    total: int = Field(ge=0)
    open: int = Field(ge=0)
    completed: int = Field(ge=0)
    deferred: int = Field(ge=0)
    not_applicable: int = Field(ge=0)
    completion_ratio: str

    @model_validator(mode="after")
    def counts_balance(self) -> ReportExecutionMetrics:
        if self.total != self.open + self.completed + self.deferred + self.not_applicable:
            raise ValueError("行动状态合计必须等于行动总数")
        return self


class ReportConsistencyCheck(BaseModel):
    code: str
    status: Literal["passed", "needs_review"]
    explanation: str


class FormalReportDocument(BaseModel):
    report_id: str
    household_id: str
    household_code: str
    household_name: str
    sequence: int = Field(ge=1)
    parent_report_id: str | None
    workflow_id: str | None
    workflow_version_id: str | None
    workflow_state: str | None
    status: ReportStatus
    title: str
    subtitle: str
    watermark: str
    chapter_count: Literal[8] = 8
    chapters: list[FormalReportChapter] = Field(min_length=8, max_length=8)
    appendices: list[ReportAppendix]
    citations: list[KnowledgeCitation]
    sourced_claims: list[ReportSourceClaim]
    numeric_ledger: list[ReportNumericClaim]
    versions: ReportVersionLedger
    execution_metrics: ReportExecutionMetrics
    consistency_checks: list[ReportConsistencyCheck]
    consistency_status: Literal["passed", "needs_review"]
    generation_trigger: ReportTrigger
    generation_reason: str
    generated_at: datetime
    analysis_date: date
    data_as_of: date
    report_hash: str
    mock_mode_supported: Literal[True] = True
    boundary_note: str

    @model_validator(mode="after")
    def enforce_structure_and_sources(self) -> FormalReportDocument:
        numbers = [chapter.number for chapter in self.chapters]
        titles = [chapter.title for chapter in self.chapters]
        if numbers != list(range(1, 9)):
            raise ValueError("正式规划书必须按 1 至 8 顺序排列")
        if tuple(titles) != FORMAL_CHAPTER_TITLES:
            raise ValueError("正式规划书一级目录必须严格保持指定八章")
        citation_ids = {item.citation_id for item in self.citations}
        referenced = {
            citation_id
            for chapter in self.chapters
            for section in chapter.sections
            for citation_id in section.citation_ids
        }
        referenced.update(
            citation_id for claim in self.sourced_claims for citation_id in claim.citation_ids
        )
        if not referenced <= citation_ids:
            raise ValueError("报告包含无法追溯的引用编号")
        if len(citation_ids) != len(self.citations):
            raise ValueError("报告引用编号不得重复")
        return self


class FormalReportSummary(BaseModel):
    report_id: str
    household_id: str
    sequence: int
    report_version: str
    parent_report_id: str | None
    workflow_id: str | None
    workflow_version_id: str | None
    status: ReportStatus
    consistency_status: Literal["passed", "needs_review"]
    generation_trigger: ReportTrigger
    generated_at: datetime
    data_as_of: date
    report_hash: str
    watermark: str
    html_url: str
    pdf_url: str


class ReportGenerationRequest(BaseModel):
    analysis_date: date | None = None
    trigger: ReportTrigger = "manual"
    reason: str = Field(default="一键生成正式八章规划书", min_length=2, max_length=500)
    expected_report_sequence: int | None = Field(default=None, ge=1)


class ReportRecalculationRequest(BaseModel):
    analysis_date: date | None = None
    trigger: Literal["monthly_review", "major_event"]
    reason: str = Field(min_length=2, max_length=500)
    expected_report_sequence: int = Field(ge=1)


class ReportActionOut(BaseModel):
    id: str
    action_code: str
    group_code: str
    title: str
    detail: str
    why: str
    completion_criteria: str
    review_cycle: str
    amount: str
    due_date: date | None
    priority: int
    status: ReportActionStatus
    completed_at: datetime | None
    deferred_until: date | None
    status_reason: str | None
    record_version: int
    calculation_source: str


class ReportActionList(BaseModel):
    household_id: str
    report_id: str | None
    report_sequence: int | None
    items: list[ReportActionOut]
    metrics: ReportExecutionMetrics


class ReportActionUpdateRequest(BaseModel):
    status: ReportActionStatus
    expected_version: int = Field(ge=1)
    reason: str = Field(min_length=2, max_length=500)
    deferred_until: date | None = None

    @model_validator(mode="after")
    def validate_deferred_date(self) -> ReportActionUpdateRequest:
        if self.status == "deferred" and self.deferred_until is None:
            raise ValueError("延期状态必须提供新的复盘日期")
        if self.status != "deferred" and self.deferred_until is not None:
            raise ValueError("只有延期状态可以提供 deferred_until")
        return self


class ReportActionUpdateResponse(BaseModel):
    action: ReportActionOut
    metrics: ReportExecutionMetrics
    report: FormalReportSummary


class ReportGenerationChain(BaseModel):
    household_id: str
    current_report_id: str | None
    chain_verified: bool
    items: list[FormalReportSummary]
    audit_event_ids: list[str]
    boundary_note: str

File: frontend/src/api/formalReport.ts
import { actorHeaders, type DemoActor } from "./actor";
import type { KnowledgeCitation } from "./trust";

export const formalChapterTitles = [
  "家庭基础情况",
  "理财目标",
  "大额支出计划",
  "理财假设",
  "家庭财务报表",
  "家庭财务比率分析",
  "投资规划建议",
  "免责声明",
] as const;

export type ReportStatus = "draft_requires_human_review" | "workflow_linked" | "client_ready";
export type ReportTrigger = "manual" | "monthly_review" | "major_event" | "action_status_change";
export type ReportActionStatus = "open" | "completed" | "deferred" | "not_applicable";

export interface ReportTable {
  title: string;
  columns: string[];
  rows: string[][];
  note: string;
  calculation_source: string;
}

export interface ReportAdvice {
  code: string;
  title: string;
  reason: string;
  priority: number;
  action: string;
  completion_criteria: string;
  review_cycle: string;
  status: ReportActionStatus;
  due_date: string | null;
  citation_ids: string[];
  calculation_source: string;
}

export interface FormalReportSection {
  code: string;
  title: string;
  narratives: string[];
  tables: ReportTable[];
  advice: ReportAdvice[];
  citation_ids: string[];
}

export interface FormalReportChapter {
  number: number;
  title: string;
  summary: string;
  sections: FormalReportSection[];
}

export interface ReportExecutionMetrics {
  total: number;
  open: number;
  completed: number;
  deferred: number;
  not_applicable: number;
  completion_ratio: string;
}

export interface FormalReport {
  report_id: string;
  household_id: string;
  household_code: string;
  household_name: string;
  sequence: number;
  parent_report_id: string | null;
  workflow_id: string | null;
  workflow_version_id: string | null;
  workflow_state: string | null;
  status: ReportStatus;
  title: string;
  subtitle: string;
  watermark: string;
  chapter_count: 8;
  chapters: FormalReportChapter[];
  appendices: Array<{ code: string; title: string; tables: ReportTable[]; narratives: string[] }>;
  citations: KnowledgeCitation[];
  sourced_claims: Array<{ text: string; citation_ids: string[] }>;
  numeric_ledger: Array<{
    code: string;
    label: string;
    raw_value: string;
    display_value: string;
    unit: string;
    source_path: string;
    calculation_source: string;
  }>;
  versions: {
    report_version: string;
    input_version: string;
    formula_version: string;
    planning_rule_version: string;
    portfolio_rule_version: string;
    twin_result_version: string;
    model_version: string;
    prompt_version: string;
    knowledge_version: string;
    product_catalog_version: string;
    fund_advisory_catalog_version?: string | null;
    workflow_version: string | null;
  };
  execution_metrics: ReportExecutionMetrics;
  consistency_checks: Array<{ code: string; status: "passed" | "needs_review"; explanation: string }>;
  consistency_status: "passed" | "needs_review";
  generation_trigger: ReportTrigger;
  generation_reason: string;
  generated_at: string;
  analysis_date: string;
  data_as_of: string;
  report_hash: string;
  mock_mode_supported: true;
  boundary_note: string;
}

export interface FormalReportSummary {
  report_id: string;
  household_id: string;
  sequence: number;
  report_version: string;
  parent_report_id: string | null;
  workflow_id: string | null;
  workflow_version_id: string | null;
  status: ReportStatus;
  consistency_status: "passed" | "needs_review";
  generation_trigger: ReportTrigger;
  generated_at: string;
  data_as_of: string;
  report_hash: string;
  watermark: string;
  html_url: string;
  pdf_url: string;
}

export interface ReportAction {
  id: string;
  action_code: string;
  group_code: string;
  title: string;
  detail: string;
  why: string;
  completion_criteria: string;
  review_cycle: string;
  amount: string;
  due_date: string | null;
  priority: number;
  status: ReportActionStatus;
  completed_at: string | null;
  deferred_until: string | null;
  status_reason: string | null;
  record_version: number;
  calculation_source: string;
}

export interface ReportActionList {
  household_id: string;
  report_id: string | null;
  report_sequence: number | null;
  items: ReportAction[];
  metrics: ReportExecutionMetrics;
}

export interface ReportGenerationChain {
  household_id: string;
  current_report_id: string | null;
  chain_verified: boolean;
  items: FormalReportSummary[];
  audit_event_ids: string[];
  boundary_note: string;
}

interface ErrorEnvelope {
  error?: { code?: string; message?: string; details?: unknown };
}

export class ReportApiError extends Error {
  readonly status: number;
  readonly code: string;

  constructor(status: number, payload: ErrorEnvelope, path: string) {
    super(payload.error?.message ?? `API ${status}: ${path}`);
    this.name = "ReportApiError";
    this.status = status;
    this.code = payload.error?.code ?? "report_api_error";
  }
}

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function readJson<T>(
  path: string,
  actor: DemoActor,
  signal?: AbortSignal,
  init?: RequestInit,
): Promise<T> {
  const response = await fetch(apiBase + path, {
    ...init,
    signal,
    headers: { Accept: "application/json", ...actorHeaders(actor), ...init?.headers },
  });
  if (!response.ok) {
    const payload = await response.json().catch(() => ({})) as ErrorEnvelope;
    throw new ReportApiError(response.status, payload, path);
  }
  return await response.json() as T;
}

export function fetchCurrentFormalReport(
  householdId: string,
  actor: DemoActor,
  signal?: AbortSignal,
): Promise<FormalReport | null> {
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/reports/current`, actor, signal);
}

export function generateFormalReport(
  householdId: string,
  actor: DemoActor,
  analysisDate: string,
): Promise<FormalReport> {
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/reports`, actor, undefined, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      analysis_date: analysisDate,
      trigger: "manual",
      reason: "客户端一键生成完整八章家庭财富规划书",
    }),
  });
}

export function recalculateFormalReport(
  householdId: string,
  actor: DemoActor,
  report: FormalReport,
  trigger: "monthly_review" | "major_event",
): Promise<FormalReport> {
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/reports/recalculate`, actor, undefined, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      analysis_date: report.analysis_date,
      trigger,
      reason: trigger === "monthly_review" ? "月度家庭事实与行动复盘" : "家庭重大事件后创建新快照并重算",
      expected_report_sequence: report.sequence,
    }),
  });
}

export function fetchReportActions(
  householdId: string,
  actor: DemoActor,
  signal?: AbortSignal,
): Promise<ReportActionList> {
  return readJson(`/api/v1/households/${encodeURIComponent(householdId)}/report-actions`, actor, signal);
}

export function updateReportAction(
  householdId: string,
  action: ReportAction,
  status: ReportActionStatus,
  actor: DemoActor,
  deferredUntil?: string,
  reason?: string,
): Promise<{ action: ReportAction; metrics: ReportExecutionMetrics; report: FormalReportSummary }> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/report-actions/${encodeURIComponent(action.action_code)}`,
    actor,
    undefined,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        status,
        expected_version: action.record_version,
        reason: reason ?? `客户将“${action.title}”标记为 ${status}`,
        deferred_until: status === "deferred" ? deferredUntil : undefined,
      }),
    },
  );
}

export function fetchReportGenerationChain(
  householdId: string,
  actor: DemoActor,
  signal?: AbortSignal,
): Promise<ReportGenerationChain> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/reports/generation-chain`,
    actor,
    signal,
  );
}

export function reportExportUrl(reportId: string, format: "html" | "pdf"): string {
  return `${apiBase}/api/v1/reports/${encodeURIComponent(reportId)}/${format}`;
}

export async function downloadFormalReport(
  report: Pick<FormalReport, "report_id" | "household_code" | "sequence">,
  format: "html" | "pdf",
  actor: DemoActor,
): Promise<void> {
  const path = reportExportUrl(report.report_id, format);
  const response = await fetch(path, { headers: actorHeaders(actor) });
  if (!response.ok) {
    const payload = await response.json().catch(() => ({})) as ErrorEnvelope;
    throw new ReportApiError(response.status, payload, path);
  }
  const objectUrl = URL.createObjectURL(await response.blob());
  const anchor = document.createElement("a");
  anchor.href = objectUrl;
  anchor.download = `wealthtwin-${report.household_code.toLowerCase()}-report-r${report.sequence}.${format}`;
  document.body.append(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(objectUrl);
}
